### DISCUSSION TITLE
rebalancing a swarm in aws auto-scaling groups

### DISCUSSION THREAD

**Discussion by nunoatgithub** [BretFisher/ama#147] [Q&A]
Hey Bret, first of all, thanks for the generous availability to receive my question :)

The question itself is:
How to maintain a swarm automatically in an aws auto-scaling group. Re-balance is not automatic as far as I know, so the new instances will be empty, even if I automate their swarm membership (hum...even if determining whether to start them as master or worker is yet another problem...).

Thank you very much in advance.
Nuno

**Comment by BretFisher** [comment]
A single `docker service update --force` will re-balance a service, but it doesn't answer your larger question.

1st, I'll assume since you're concerned with auto-scaling and ASG's that you have a decent size Swarm (a dozen instances) with potential capacity issues.

2nd, If you're even thinking you're big enough for ASG's, your managers should be in their own ASG that's set to 3 or 5 instances and not mixed with your workers. cloud-init should have that ASG auto-join as managers and your user-land services should all use placement constraints to keep them on workers. Leave managers for admin, monitoring, etc.

3rd I'm going to suggest using AWS features here since that's what you mentioned. If you wanted to do everything you could inside the Swarm, you could try plugging in Prometheus and InfraKit for a self-aware self-healing Swarm, but I'll 

Problem 1: Many containers on many servers means you have a nested performance problem. When you had just one-app-per-VM in ASG's, it was an easy linear up and down scaling issue. If you have 5 containers per server (a small number for most setups), how do you know which one should be "moved" or have its replicas increased to a new node that just spun up?

Problem 2: You might have multiple types of workloads in your Swarm. If your ASG scales down, those workloads will pop up somewhere else. You're scaling has to be smart and conservative so it knows to leave enough room for container movement and growth on existing servers. If you have anything in your Swarm with persistent data or sensitive connections, like database API's, you may not want those killed and spun back up on various nodes all the time just so ASG's can "right size" your set of servers.

Bottom line is ASG's alone can't handle this complexity.

Option 1: one container per VM: this isn't ideal, but in cases where you know you're going to have capacity issues on a Swarm service, give that service (or set of services) its own ASG, have the instances spin up and auto-join the swarm as workers with a specific host label (labels that go in /etc/docker/daemon.json) and then have the service constrained to only nodes with that label, in Global Mode. Then when a new instance of that ASG is added, it auto-increases the task count to match the number of instances. This is the easy way because you can then usually pick the appropriate EC2 instance for that custom ASG that best fits those workloads.

Option 2: Have all your Swarm workers in a single ASG that autoscales *up* on performance alerts. You can try autoscale down but this can be tricky based on what I said above. It just depends on your workloads. Then, implement advanced service/task monitoring like Prometheus or any Swarm-aware solution, and when individual tasks hit CPU limits, you could have it kick off a `docker service replicas...` to scale up the service. Now, that won't increase node count, so you'll have two different unaware monitoring solutions dealing with two layers of your solution. Not ideal, but usable in the right scenario. 

Also, you'll likely want a `docker service update --force` on all affected services when the ASG adds a node in this case, and it can't be done from the new node itself because it's not a manager.  This is where [CloudWatch Events](https://docs.aws.amazon.com/autoscaling/ec2/userguide/cloud-watch-events.html) come into play. You could have a ASG Instance Launch event kick off something that would tell the Docker API to increase replicas on specific services.

Lastly, I don't pretend to know all the projects going on there around auto-scaling services and instances, so if you see anything useful for Swarm, be sure to add it to this thread :)

**Comment by nunoatgithub** [comment]
What a wonderfully detailed and helpful answer Bret !! Thank you so much!

I am definitely always considering the case of one container per ec2 instance, otherwise it becomes a bit hellish to control, as you so well underline. 

To be honest, my interest about ASGs is even more related with availability than capacity, which means that in certain cases I would gladly consider a cluster as little as just three nodes, all of them having to be managers, due to the manager's quorum constraints (as far as I understood them). I love the host labels solution you describe and will definitely keep them in mind, but in this case one alternative that came to mind after reading your answer, could perhaps be to just have a boot script in the new node that would set it as a manager (to have access to the swarm api), and from that node evaluate the number of already existent managers and automatically de-promote itself to worker in case enough existed already ? Then, if state-per-container would not be an issue, I would do the automatic **docker service update --force** as you describe it in your next-to-last paragraph. 
If state-per-node would be an issue however, my understanding is that there is nothing swarm can do about it and the app itself needs to complement the re-balancing by managing its graceful shutdown somehow (maybe like indicated in this [link](https://sreeninet.wordpress.com/2017/08/15/docker-features-for-handling-containers-death-and-resurrection/))

The likelihood of quorum loss in cases where two of the nodes of a three-node cluster would be momentarily lost (before AS kicks in and saves the day) would also require special measures I guess.

Due to the complexity of these scenarios, I'm looking very favorably now to the Fargate AWS product and considering dropping swarm altogether.

**Comment by BretFisher** [comment]
Yea "managed container platforms" will always be easier than deploying your own container solution on bare OS's. Remember that Docker was created by a company trying to compete with Heroku, and ultimately open sourced how they built the App PaaS into what became Docker. :)

The one concern with your idea on "temp manager during startup" would be that most of the time you'll have your workers in the dmz and managers more protected and not internet accessible. In that case you're creating Swarm services that are "worker only" by using node labels and service constraints. In this case your idea wouldn't work, as the update --force wouldn't let any of the services spread to the new node until it was demoted. But it's still some creative thinking.
